<!-- rosrun gazebo_ros spawn_model -model F-150 -sdf -file /home/bigb/gazebo_f150_ws/src/gazebo_f150/models/F150/model.sdf -->


<?xml version="1.0" ?>
<sdf version="1.5">
  <model name="box">
    <pose>0 0 0 0 0 0</pose>
    <static>false</static>

    <link name="chassis">
    <pose>0 0 0 0 0 0</pose>
      <inertial>
        <inertia>
          <ixx>1000</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>1000</iyy>
          <iyz>0</iyz>
          <izz>1000</izz>
        </inertia>
        <mass>1000</mass>
      </inertial>
      <collision name="base_link_collision">
        <pose>0 0 0.3 0 0 0</pose>
        <geometry>
          <box>
            <size>0.1 0.1 0.1</size>
          </box>
        </geometry>
      </collision>
      <visual name="base_link_visual">
        <geometry>
          <cylinder>
            <radius>0.1</radius>
            <length>1.0</length>
          </cylinder>
        </geometry>
      </visual>
    </link>

    <link name="steering_centerlink_link">
    <pose>1.75 0 0.445 0 0 0</pose>
      <inertial>
        <inertia>
          <ixx>0.1</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.1</iyy>
          <iyz>0</iyz>
          <izz>0.1</izz>
        </inertia>
        <mass>0.1</mass>
      </inertial>
      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
            <box>
              <size>0.1 0.1 0.1</size>
            </box>
        </geometry>
      </collision>
      <visual name="visual">
        <pose>0 0 0 1.57079632679 0 0</pose>
        <geometry>
          <cylinder>
            <radius>0.01</radius>
            <length>0.5</length>
          </cylinder>
        </geometry>
      </visual>
    </link>

    <joint name="centerlink2chassis_joint" type="prismatic">
      <parent>chassis</parent>
      <child>steering_centerlink_link</child>
      <pose>0 0 0 0 0 0</pose>
      <axis>
        <xyz>0 1 0</xyz>
      </axis>
    </joint>

    <!-- Start Front Right Wheel -->
        <link name="front_right_wheel">
        <pose>1.953 -0.8665 0.445 0 0 0</pose>
          <inertial>
            <inertia>
              <ixx>2</ixx>
              <ixy>0</ixy>
              <ixz>0</ixz>
              <iyy>3.46</iyy>
              <iyz>0</iyz>
              <izz>2</izz>
            </inertia>
            <mass>35.0</mass>
          </inertial>

          <collision name="front_right_wheel_collision">
          <pose>0 0 0 1.57079633 0 0</pose>
            <geometry>
              <cylinder>
                <radius>0.445</radius>
                <length>0.297</length>
              </cylinder>
            </geometry>
            <surface>
              <friction>
                <ode>
                  <mu>0.7</mu>
                  <mu2>0.7</mu2>
                </ode>
              </friction>
            </surface>
          </collision>

          <visual name="front_right_wheel_visual">
          <pose>0 0 0 0 0 0</pose>
            <geometry>
              <mesh>
                <uri>model://F150/meshes/wheel.dae</uri>
              </mesh>
            </geometry>
          </visual>
        </link>


    <link name="right_steering_arm_link">
    <pose>1.953 -0.8665 0.445 0 0 -0.23300146</pose> <!-- Front wheel position with rotation -->
      <inertial>
        <inertia>
          <ixx>0.1</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.1</iyy>
          <iyz>0</iyz>
          <izz>0.1</izz>
        </inertia>
        <mass>0.1</mass>
      </inertial>
      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <cylinder>
            <radius>0.005</radius>
            <length>0.01</length>
          </cylinder>
        </geometry>
      </collision>
      <visual name="visual">
        <pose>-0.125 0 0 0 1.57079632679 0</pose>
        <geometry>
          <cylinder>
            <radius>0.01</radius>
            <length>0.25</length>
          </cylinder>
        </geometry>
      </visual>
    </link>

    <link name="right_tierod_link">
    <pose>1.953 -0.8665 0.445 0 0 -0.23300146</pose> <!-- Front wheel position with rotation -->
      <inertial>
        <inertia>
          <ixx>0.1</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.1</iyy>
          <iyz>0</iyz>
          <izz>0.1</izz>
        </inertia>
        <mass>0.1</mass>
      </inertial>
      <collision name="collision">
        <pose>0 0 0 0 0 0</pose>
        <geometry>
          <box>
            <size>0.1 0.1 0.1</size>
          </box>
        </geometry>
      </collision>
      <visual name="visual">
        <pose>-0.375 0 0 0 1.57079632679 0</pose>
        <geometry>
          <cylinder>
            <radius>0.01</radius>
            <length>0.25</length>
          </cylinder>
        </geometry>
      </visual>
    </link>

    <joint name="right_arm2wheel_joint" type="revolute">
      <parent>right_steering_arm_link</parent>
      <child>front_right_wheel</child>
      <pose>0 0 0 0 0 0</pose>
      <axis>
        <xyz>0 1 0</xyz>
      </axis>
    </joint>

    <joint name="right_arm2tierod_joint" type="ball">
      <parent>right_tierod_link</parent>
      <child>right_steering_arm_link</child>
      <pose>-0.25 0 0 0 0 0</pose>
    </joint>
<!--
    <joint name="right_tierod2centerlink_joint" type="ball">
      <parent>steering_centerlink_link</parent>
      <child>right_tierod_link</child>
      <pose>-0.5 0 0 0 0 0</pose>
    </joint>
-->
    <joint name="right_wheel2chassis_joint" type="revolute">
      <parent>right_steering_arm_link</parent>
      <child>chassis</child>
      <pose>1.953 -0.8665 0.445 0 0 0</pose> <!-- Front wheel position -->
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>

    <!-- END right steering -->
    <!-- End Front Right Wheel -->


    <!-- Front Left Wheel -->
        <link name="front_left_wheel">
        <pose>1.953 0.8665 0.445 0 0 0</pose>
          <inertial>
            <inertia>
              <ixx>2</ixx>
              <ixy>0</ixy>
              <ixz>0</ixz>
              <iyy>3.46</iyy>
              <iyz>0</iyz>
              <izz>2</izz>
            </inertia>
            <mass>35.0</mass>
          </inertial>

          <collision name="front_left_wheel_collision">
          <pose>0 0 0 1.57079633 0 0</pose>
            <geometry>
              <cylinder>
                <radius>0.445</radius>
                <length>0.297</length>
              </cylinder>
            </geometry>
            <surface>
              <friction>
                <ode>
                  <mu>0.7</mu>
                  <mu2>0.7</mu2>
                </ode>
              </friction>
            </surface>
          </collision>

          <visual name="front_left_wheel_visual">
          <pose>0 0 0 0 0 3.1416</pose>
            <geometry>
              <mesh>
                <uri>model://F150/meshes/wheel.dae</uri>
              </mesh>
            </geometry>
          </visual>
        </link>


            <link name="left_steering_arm_link">
            <pose>1.953 0.8665 0.445 0 0 0.23300146</pose> <!-- Front wheel position with rotation -->
              <inertial>
                <inertia>
                  <ixx>0.1</ixx>
                  <ixy>0</ixy>
                  <ixz>0</ixz>
                  <iyy>0.1</iyy>
                  <iyz>0</iyz>
                  <izz>0.1</izz>
                </inertia>
                <mass>0.1</mass>
              </inertial>
              <collision name="collision">
                <pose>0 0 0 0 0 0</pose>
                <geometry>
                  <cylinder>
                    <radius>0.005</radius>
                    <length>0.01</length>
                  </cylinder>
                </geometry>
              </collision>
              <visual name="visual">
                <pose>-0.125 0 0 0 1.57079632679 0</pose>
                <geometry>
                  <cylinder>
                    <radius>0.01</radius>
                    <length>0.25</length>
                  </cylinder>
                </geometry>
              </visual>
            </link>

            <link name="left_tierod_link">
            <pose>1.953 0.8665 0.445 0 0 0.23300146</pose> <!-- Front wheel position with rotation -->
              <inertial>
                <inertia>
                  <ixx>0.1</ixx>
                  <ixy>0</ixy>
                  <ixz>0</ixz>
                  <iyy>0.1</iyy>
                  <iyz>0</iyz>
                  <izz>0.1</izz>
                </inertia>
                <mass>0.1</mass>
              </inertial>
              <collision name="collision">
                <pose>0 0 0 0 0 0</pose>
                <geometry>
                  <box>
                    <size>0.1 0.1 0.1</size>
                  </box>
                </geometry>
              </collision>
              <visual name="visual">
                <pose>-0.375 0 0 0 1.57079632679 0</pose>
                <geometry>
                  <cylinder>
                    <radius>0.01</radius>
                    <length>0.25</length>
                  </cylinder>
                </geometry>
              </visual>
            </link>

            <joint name="left_arm2wheel_joint" type="revolute">
              <parent>left_steering_arm_link</parent>
              <child>front_left_wheel</child>
              <pose>0 0 0 0 0 0</pose>
              <axis>
                <xyz>0 1 0</xyz>
              </axis>
            </joint>

            <joint name="left_arm2tierod_joint" type="ball">
              <parent>left_tierod_link</parent>
              <child>left_steering_arm_link</child>
              <pose>-0.25 0 0 0 0 0</pose>
            </joint>

            <joint name="left_tierod2centerlink_joint" type="ball">
              <parent>steering_centerlink_link</parent>
              <child>left_tierod_link</child>
              <pose>-0.5 0 0 0 0 0</pose>
            </joint>

            <joint name="left_wheel2chassis_joint" type="revolute">
              <parent>left_steering_arm_link</parent>
              <child>chassis</child>
              <pose>1.953 0.8665 0.445 0 0 0</pose> <!-- Front wheel position -->
              <axis>
                <xyz>0 0 1</xyz>
              </axis>
            </joint>
    <!-- End Front Left Wheel -->

    <!-- Start Rear Right Wheel -->
        <link name="rear_right_wheel">
        <pose>-1.698 -0.8665 0.445 0 0 0</pose>
          <inertial>
            <inertia>
              <ixx>2</ixx>
              <ixy>0</ixy>
              <ixz>0</ixz>
              <iyy>3.46</iyy>
              <iyz>0</iyz>
              <izz>2</izz>
            </inertia>
            <mass>35.0</mass>
          </inertial>

          <collision name="rear_right_wheel_collision">
          <pose>0 0 0 1.57079633 0 0</pose>
            <geometry>
              <cylinder>
                <radius>0.445</radius>
                <length>0.297</length>
              </cylinder>
            </geometry>
            <surface>
              <friction>
                <ode>
                  <mu>0.7</mu>
                  <mu2>0.7</mu2>
                </ode>
              </friction>
            </surface>
          </collision>

          <visual name="rear_right_wheel_visual">
          <pose>0 0 0 0 0 0</pose>
            <geometry>
              <mesh>
                <uri>model://F150/meshes/wheel.dae</uri>
              </mesh>
            </geometry>
          </visual>
        </link>

        <joint name="rear_right_wheel_joint" type="revolute">
          <parent>chassis</parent>
          <child>rear_right_wheel</child>
          <pose>0 0 0 0 0 0</pose>
          <axis>
            <xyz>0 1 0</xyz>
          </axis>
        </joint>
    <!-- End Rear Right Wheel -->

    <!-- Start Rear Left Wheel -->
        <link name="rear_left_wheel">
        <pose>-1.698 0.8665 0.445 0 0 0</pose>
          <inertial>
            <inertia>
              <ixx>2</ixx>
              <ixy>0</ixy>
              <ixz>0</ixz>
              <iyy>3.46</iyy>
              <iyz>0</iyz>
              <izz>2</izz>
            </inertia>
            <mass>35.0</mass>
          </inertial>

          <collision name="rear_left_wheel_collision">
          <pose>0 0 0 1.57079633 0 0</pose>
            <geometry>
              <cylinder>
                <radius>0.445</radius>
                <length>0.297</length>
              </cylinder>
            </geometry>
            <surface>
              <friction>
                <ode>
                  <mu>0.7</mu>
                  <mu2>0.7</mu2>
                </ode>
              </friction>
            </surface>
          </collision>

          <visual name="rear_left_wheel_visual">
          <pose>0 0 0 0 0 3.1416</pose>
            <geometry>
              <mesh>
                <uri>model://F150/meshes/wheel.dae</uri>
              </mesh>
            </geometry>
          </visual>
        </link>

        <joint name="rear_left_wheel_joint" type="revolute">
          <parent>chassis</parent>
          <child>rear_left_wheel</child>
          <pose>0 0 0 0 0 0</pose>
          <axis>
            <xyz>0 1 0</xyz>
          </axis>
        </joint>
    <!-- End Rear Left Wheel -->


</model>
</sdf>
